File: python/BinarySearchTreetoGreaterSumTree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def bstToGst(root):
    # Get the node with the smallest value greater than this one.
    def get_successor(node):
        succ = node.right
        while succ.left is not None and succ.left is not node:
            succ = succ.left
        return succ

    total = 0
    node = root
    while node is not None:
        # If there is no right subtree, then we can visit this node and
        # continue traversing left.
        if node.right is None:
            total += node.val
            node.val = total
            node = node.left
        # If there is a right subtree, then there is a node that has a
        # greater value than the current one. therefore, we must traverse
        # that node first.
        else:
            succ = get_successor(node)
            # If there is no left subtree (or right subtree, because we are
            # in this branch of control flow), make a temporary connection
            # back to the current node.
            if succ.left is None:
                succ.left = node
                node = node.right
            # If there is a left subtree, it is a link that we created on
            # a previous pass, so we should unlink it and visit this node.
            else:
                succ.left = None
                total += node.val
                node.val = total
                node = node.left

    return root

# Helper function to construct a tree from list representation
def construct_tree(nodes):
    if not nodes:
        return None
    idx = 0
    root = TreeNode(nodes[idx])
    idx += 1
    queue = [root]
    while queue:
        node = queue.pop(0)
        if idx < len(nodes) and nodes[idx] is not None:
            node.left = TreeNode(nodes[idx])
            queue.append(node.left)
        idx += 1
        if idx < len(nodes) and nodes[idx] is not None:
            node.right = TreeNode(nodes[idx])
            queue.append(node.right)
        idx += 1
    return root

# Helper function to convert tree to list representation
def tree_to_list(root):
    if not root:
        return []
    result = []
    queue = [root]
    while queue:
        node = queue.pop(0)
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result

# Example usage
input_tree = [4,1,6,0,2,5,7,None,None,None,3,None,None,None,8]
root = construct_tree(input_tree)
bstToGst(root)
output_list = tree_to_list(root)
print(output_list)  # Output: [30,36,21,36,35,26,15,None,None,None,33,None,None,None,8]

File: JAVA/java/src/BinarySearchTreetoGreaterSumTree.java
// Definition for a binary tree node.

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class BinarySearchTreetoGreaterSumTree {

    // Method to transform BST to GST
    public TreeNode bstToGst(TreeNode root) {
        // Get the node with the smallest value greater than this one.
        TreeNode[] succ = new TreeNode[1];
        succ[0] = null;
        return bstToGstHelper(root, succ);
    }

    // Helper method for recursion
    private TreeNode bstToGstHelper(TreeNode root, TreeNode[] succ) {
        if (root == null) {
            return null;
        }

        // Recursively traverse the right subtree
        root.right = bstToGstHelper(root.right, succ);

        // Update root's value with accumulated sum
        if (succ[0] != null) {
            root.val += succ[0].val;
        }

        // Update the successor node
        succ[0] = root;

        // Recursively traverse the left subtree
        root.left = bstToGstHelper(root.left, succ);

        return root;
    }

    // Method to convert BST to level-order list format
    public List<Integer> bstToLevelOrder(TreeNode root) {
        List<Integer> levelOrderList = new ArrayList<>();
        if (root == null) {
            return levelOrderList;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            if (current != null) {
                levelOrderList.add(current.val);
                queue.offer(current.left);
                queue.offer(current.right);
            } else {
                levelOrderList.add(null);
            }
        }

        // Remove trailing nulls
        while (!levelOrderList.isEmpty() && levelOrderList.get(levelOrderList.size() - 1) == null) {
            levelOrderList.remove(levelOrderList.size() - 1);
        }

        return levelOrderList;
    }

    // Helper method to construct a tree from array representation
    public TreeNode constructTree(Integer[] nodes) {
        if (nodes == null || nodes.length == 0) {
            return null;
        }
        return constructTreeHelper(nodes, 0);
    }

    private TreeNode constructTreeHelper(Integer[] nodes, int index) {
        TreeNode root = null;
        if (index < nodes.length && nodes[index] != null) {
            root = new TreeNode(nodes[index]);
            root.left = constructTreeHelper(nodes, 2 * index + 1);
            root.right = constructTreeHelper(nodes, 2 * index + 2);
        }
        return root;
    }

    // Main method for testing
    public static void main(String[] args) {
        BinarySearchTreetoGreaterSumTree solution = new BinarySearchTreetoGreaterSumTree();

        // Example input: [4,1,6,0,2,5,7,null,null,null,3,null,null,null,8]
        Integer[] input = {4,1,6,0,2,5,7,null,null,null,3,null,null,null,8};
        TreeNode root = solution.constructTree(input);

        // Transform BST to GST
        TreeNode gstRoot = solution.bstToGst(root);

        // Convert GST to level-order list format
        List<Integer> levelOrder = solution.bstToLevelOrder(gstRoot);

        // Print level-order list
        System.out.println(levelOrder);  // Output: [30, 36, 21, 36, 35, 26, 15, null, null, null, 33, null, null, null, 8]
    }
}

